proposals/0000-visible-forall.rst

```diff
@@ -0,0 +1,1269 @@
+   * Constraint arrows: ``a => b``
+   * Universal quantification: ``forall a. b``
+   * Visible universal quantification: ``forall a -> b``.
+
+   We will call them **types-in-terms**.
+
+   Grammatically, their constituents are terms, not types::
+
+                   proposed grammar:                      as opposed to:
+         ┌────────────────────────────────────┬───────────────────────────────────────┐
+         │                                    │                                       │
+         │  exp ::=                           │    exp ::=                            │
+         │      | exp₀ '->' exp₁              │        | type₀ '->' type₁             │
+         │      | exp₀ '=>' exp₁              │        | type₀ '=>' type₁             │
+         │      | 'forall' tv_bndrs '.'  exp  │        | 'forall' tv_bndrs '.'  type  │
+         │      | 'forall' tv_bndrs '->' exp  │        | 'forall' tv_bndrs '->' type  │
+         │                                    │                                       │
+         └────────────────────────────────────┴───────────────────────────────────────┘
+
+   This is a necessity to avoid parsing conflicts, with the following
+   consequences:
+
+   1. The ``'`` symbol signifies Template Haskell name quotation rather than ``DataKinds`` promotion.
+   2. The ``*`` symbol is treated as an infix operator regardless of ``-XStarIsType``.
+   3. Built-in syntax for tuples and lists is interpreted as in terms.
+      That is, ``[a]`` is a singleton list rather than the type of a list,
+      and ``(a, b)`` is a pair rather than the type of a pair.
+
+2. Make ``forall`` a keyword at the term level. Not guarded by any extension
+   (same motivation as `#193 <https://github.com/ghc-proposals/ghc-proposals/blob/master/proposals/0193-forall-keyword.rst>`_).
+   This implies ``forall`` is no longer a valid identifier.
+
+3. Introduce a new extension, ``ListTupleTypeSyntax``, on by default,
+   which enables:
+
+   * ``[]`` as the list type constructor
+   * ``()`` as the unit type
+   * ``[a]`` the syntax of a list type
+   * ``(,)`` as the pair type constructor
+   * ``(a, b)`` as the syntax of a pair type
+
+   When the extension is on, these constructs retain their Haskell 2010
+   meaning, which depends on whether we are in a type-level or term-level
+   context.
+
+   When the extension is off, all of the above are interpreted as in terms:
+
+   * ``[]`` is always an empty list
+   * ``()`` is always the unit data constructor
+   * ``[a]`` is always a singleton list (not the list type)
+   * ``(,)`` is always the pair data constructor (not the type constructor)
+   * ``(a, b)`` is always a pair (not the type of a pair)
+
+   Likewise for tuples of higher arities.
+
+   Export the following synonym from the ``Data.List`` module::
+
+     type List = []
+
+   Export the following synonyms from the ``Data.Tuple`` module::
+
+     type Unit = ()
```

Comment:
Why not `TupleOf0`? And should we include `TupleOf1 = Solo`?

Reply:
Because I prefer human-readable names.
1. `Unit` because `()` is commonly called “the unit type”
2. No `TupleOf1` there’s already `Solo` and it’s a better name

The goal here is to avoid punning, not to create `TupleOfN` names. I would also say `Pair` and `Triple` instead of `(,)` and `(,,)`, but I’m not sure if `Quadruple` is a good idea and it feels strange to stop at 3.